### PR TITLE
bazel: migrate to bazelmod and bazel 7

### PR DESCRIPTION
**Pull request by lromor**

### CODE REVIEW

**Comment by corco** [comment]
Couldn't we create a WORKSPACE.bzlmod and keep WORKSPACE as it were to keep compatibility with bazel 5?

**Comment by lromor** [comment]
@corco , I didn't know you could do that! Nice! I will try to add those changes as soon as I have some time.

**Comment by hzeller** [comment]
Question to the experts: What is the difference between `MODULES.bazel` vs `WORKSPACE.bzlmod` ? I have seen `MODULES.bazel` in other bzlmodded projects.

**Review by hzeller — COMMENTED** [review]
Nice!
Mostly some questions for now below.

Would you like to add one simple addition to the CI that just does a compile and test using bazel-7 with bzlmod enabled ?

**Comment by corco** [comment]
> Question to the experts: What is the difference between `MODULES.bazel` vs `WORKSPACE.bzlmod` ? I have seen `MODULES.bazel` in other bzlmodded projects.

`WORKSPACE.bzlmod` is used in place of `WORKSPACE` when `--enable_bzlmod` options is active. It should eventually become empty once everything is moved to bzlmod, and removed once "non-bzlmod" build are not supported anymore

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+anytree==2.8.0
```
Comment (hzeller) [inline requirements.txt:1]: This file is not needed anymore, right ?

File: shell.nix [hunk 2]
```diff
@@ -12,7 +12,7 @@ verible_used_stdenv.mkDerivation {
   name = "verible-build-environment";
   buildInputs = with pkgs;
     [
-      bazel_5
+      bazel_7
       jdk11
```
Comment (hzeller) [inline shell.nix:16]: Maybe change this to bazel 6 as a good middle-ground.

That way, the user can test the 'old' form (without `--enable_bzlmod`) and the new.

**Comment by hzeller** [comment]
Looks good. If you think it is ready to submit @lromor , remove the 'draft' status of this PR.

Adding a bazel-7 CI test then would be a good follow-up PR.

**Comment by hzeller** [comment]
`.github/bin/make-compilation-db.sh` stopped working with bazel 7, but I think we can address that later.

```
ERROR: Unable to find package for @@[unknown repo 'bazel_compdb' requested from @@]//:aspects.bzl: The repository '@@[unknown repo 'bazel_compdb' requested from @@]' could not be resolved: No repository visible as '@bazel_compdb' from main repository.
ERROR: Analysis of aspects '[@@[unknown repo 'bazel_compdb' requested from @@]//:aspects.bzl%compilation_database_aspect] with parameters {} on //common/analysis:command-file-lexer' failed; build aborted: Unable to find package for @@[unknown repo 'bazel_compdb' requested from @@]//:aspects.bzl: The repository '@@[unknown repo 'bazel_compdb' requested from @@]' could not be resolved: No repository visible as '@bazel_compdb' from main repository.
```


I tried the `hedron_compile_commands` before, but was not successful (`.github/bin/run-clang-tidy-cached.cc` doesn't work with it). Something to figure out while we still have bazel 5 going.

**Comment by lromor** [comment]
> `.github/bin/make-compilation-db.sh` stopped working with bazel 7, but I think we can address that later.
> 
> ```
> ERROR: Unable to find package for @@[unknown repo 'bazel_compdb' requested from @@]//:aspects.bzl: The repository '@@[unknown repo 'bazel_compdb' requested from @@]' could not be resolved: No repository visible as '@bazel_compdb' from main repository.
> ERROR: Analysis of aspects '[@@[unknown repo 'bazel_compdb' requested from @@]//:aspects.bzl%compilation_database_aspect] with parameters {} on //common/analysis:command-file-lexer' failed; build aborted: Unable to find package for @@[unknown repo 'bazel_compdb' requested from @@]//:aspects.bzl: The repository '@@[unknown repo 'bazel_compdb' requested from @@]' could not be resolved: No repository visible as '@bazel_compdb' from main repository.
> ```
> 
> I tried the `hedron_compile_commands` before, but was not successful (`.github/bin/run-clang-tidy-cached.cc` doesn't work with it). Something to figure out while we still have bazel 5 going.

Alright, then the PR is ready.

**Comment by hzeller** [comment]
I played a little bit on my machine and it seems to work except make-compile-commands with bazel 7.
@corco What do you think, anything else you see ?

**Comment by corco** [comment]
> I played a little bit on my machine and it seems to work except make-compile-commands with bazel 7.
> 
> @corco What do you think, anything else you see ?

I'm off this week. I will take a look Monday if this PR hasn't been merged.

**Comment by hzeller** [comment]
Alright, then merging it now as it seems a good starting point; even if there are issues, everything still works with older `bazel`, so they can be addressed as need be.

**Pull request merged**